### PR TITLE
Fix a bug that doesn't release lock

### PR DESCRIPTION
**Pull request by lam2003**
err!=nil时未释放锁，使用defer p.lock.Unlock()修复

---
name: Pull request
about: Propose changes to the code
title: ''
labels: ''
assignees: ''
---

## 4. Checklist

- [ ] I have squashed all insignificant commits.
- [ ] I have commented my code for explaining package types, values, functions, and non-obvious lines.
- [ ] I have written unit tests and verified that all tests passes (if needed).
- [ ] I have documented feature info on the README (only when this PR is adding a new feature).
- [x] (optional) I am willing to help maintain this change if there are issues with it later.

### CODE REVIEW

File: pool.go [hunk 1]
```diff
@@ -247,6 +247,7 @@ func (p *Pool) revertWorker(worker *goWorker) bool {
 	}
 	worker.recycleTime = time.Now()
 	p.lock.Lock()
+	defer p.lock.Unlock()
```
Comment (panjf2000) [inline pool.go]: 在这种性能极度敏感的库代码里尽量不要用 defer，它在汇编层面会有很多连锁调用导致性能损耗。直接在下面的 `if err != nil {}` 里面加一个解锁的步骤就行了。

Comment (choleraehyq) [reply pool.go]: defer 在 1.14 基本已经没啥损耗了。另外即使是在老版本，大概率池子里运行的函数本身远重于 defer。这个池子本身的性能瓶颈也不会在这里。反倒是因为不用 defer，这已经是第二次碰到异常路径上没 unlock 的问题了。

**Comment by choleraehyq** [comment]
@panjf2000 这种 case 不加单测么？

File: pool.go (hunk 1 above)
Comment (panjf2000) [reply pool.go]: 能看出具体是哪一步没 unlock 吗？

**Pull request merged**